pkg/controller/registry/resolver/instrumented_resolver.go

```diff
@@ -0,0 +1,36 @@
+package resolver
+
+import (
+	"time"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+type InstrumentedResolver struct {
+	wrapped               Resolver
+	successMetricsEmitter func(*float64)
+	failureMetricsEmitter func(*float64)
```

Comment:
Could we make this a single function which takes a boolean, or checks if the error is nil like this:
`ResolutionMetricsEmitter func(*float64, error)`

Reply:
@awgreene keeping them as two separate functions is actually preferable since it adheres to the single task per function guideline. That is evident if we think about making the corresponding change to the unit test case: 

Corresponding change required to the unit test case: 
`changeToZero(*float64)` and `changeToOne(*float64)` gets merged to `changeValue(*float64, error)`
`changeValue` then needs to be implemented, which means `ResolutionMetricsEmitter` is really not being unit tested. 

Essentially the way I've written it now gives us the most coverage with unit testing.

Reply:
I see your point. I had classified "reporting the success or failure of a resolution" as a single task for the function. I had thought that passing in an error to a function that could report the success or failure would still allow the `InstrumentedResolver` to effectively "pick" the metric to emit.

If you guys disagree, I am fine marking this as resolved.

Reply:
It's an ivory tower distinction, but deciding that a nil error means success and a non-nil error means failure is a specific way of interpreting values return from the Resolver interface's method. If that method were to change in the future -- for example, if it only returned a `bool` indicating success or failure -- now this component would either need to construct a fake non-nil error solely to indicate that the metrics package should record a failure metric, or the function that emits metrics would need to change to accept a bool instead of an error.

I didn't find any (to me) convincing essays against boolean parameters as mode switches, but I find that they usually indicate that the implementation will contain "if B then DoX() else DoY()," and are better off being two functions, so my preference is to avoid them.

https://martinfowler.com/bliki/FlagArgument.html

Reply:
Consider me convinced :+1: